### PR TITLE
feat(useOffsetPagination): new function

### PR DESCRIPTION
**Pull request by okxiaoliang4**
Composition api use Cases

```typescript
const {
  currentPage,
  currentPageSize,
  pageCount,
  isFirstPage,
  isLastPage,
  prev,
  next,
} = useOffsetPagination(
  {
    total: database.value.length,
    page: 1,
    pageSize,
    onPageChange: fetchData,
    onPageSizeChange: fetchData,
  },
)
```

Component use Cases
```vue
<UseOffsetPagination
  v-slot="{
    currentPage,
    currentPageSize,
    next,
    prev,
    pageCount,
    isFirstPage,
    isLastPage
  }"
  :total="database.length"
  @page-change="fetchData"
  @page-size-change="fetchData"
  @page-count-change="onPageCountChange"
>
  <!-- your code -->
</UseOffsetPagination>
```
or props event callback
```vue
<UseOffsetPagination
  v-slot="{
    currentPage,
    currentPageSize,
    next,
    prev,
    pageCount,
    isFirstPage,
    isLastPage
  }"
  :total="database.length"
  :on-page-change="fetchData"
  :on-page-size-change="fetchData"
  :on-page-count-change="onPageCountChange"
>
  <!-- your code -->
</UseOffsetPagination>
```

### CODE REVIEW

File: packages/core/usePagination/component.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { defineComponent, reactive } from 'vue-demi'
+import { usePagination } from '@vueuse/core'
+
+export const UsePagination = defineComponent({
```
Comment (wheatjs) [inline packages/core/usePagination/component.ts:4]: I'd prefer us to use `defineComponent<UsePaginationOptions>` for props

**Review by okxiaoliang4 — COMMENTED** [review]
self reviewed

**Comment by wheatjs** [comment]
Since there are more types of pagination out there, maybe we should call this function `useOffsetPagination`? We could also provide a function for cursor pagination as well perhaps

**Comment by okxiaoliang4** [comment]
> Since there are more types of pagination out there, maybe we should call this function `useOffsetPagination`? We could also provide a function for cursor pagination as well perhaps

I agree.

File: packages/core/useOffsetPagination/index.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import type { UnwrapNestedRefs } from 'vue-demi'
+import { computed, isRef, reactive, unref, watch } from 'vue-demi'
+import type { MaybeRef } from '@vueuse/core'
+import { biSyncRef, noop, useClamp } from '@vueuse/core'
+
+export interface UseOffsetPaginationOptions {
+  /**
+     * The number of items to show per page.
```
Comment (wheatjs) [inline packages/core/useOffsetPagination/index.ts:8]: Incorrect doc for prop

**Comment by wheatjs** [comment]
Wondering if we could make the `total` prop optional? I know some APIs don't provide a total count which might cause issues when using this utility

**Comment by okxiaoliang4** [comment]
> Wondering if we could make the `total` prop optional? I know some APIs don't provide a total count which might cause issues when using this utility

I can set `Infinity` to the default value.
If `options.total` is `undefined` then return value remove ref:`isLastPage`

Like this
```typescript
export interface UseOffsetPaginationReturn {
  currentPage: Ref<number>
  currentPageSize: Ref<number>
  pageCount: ComputedRef<number>
  isFirstPage: ComputedRef<boolean>
  isLastPage: ComputedRef<boolean>
  prev: () => void
  next: () => void
}

export type UseOffsetPaginationInfinityPageReturn = Omit<UseOffsetPaginationReturn, 'isLastPage'>

export function useOffsetPagination(options: Omit<UseOffsetPaginationOptions, 'total'>): UseOffsetPaginationInfinityPageReturn
export function useOffsetPagination(options: UseOffsetPaginationOptions): UseOffsetPaginationReturn
```

**Review by wheatjs — DISMISSED** [review]

**Comment by wheatjs** [comment]
Look good to me, maybe we could get some tests as well?

**Comment by antfu** [comment]
@okxiaoliang4 Thanks for your contributions and welcome to the team :)

**Comment by okxiaoliang4** [comment]
> @okxiaoliang4 Thanks for your contributions and welcome to the team :)
Thanks,It's my pleasure to be part of the team. (^w^)

**Comment by antfu** [comment]
@okxiaoliang4 Join the Discord and ping me :) https://chat.antfu.me/

**Comment by yeyan1996** [comment]
Is it possible to add a loading state？

**Pull request merged**